### PR TITLE
Fixing the internal state management, reverting actions to take references to state data, cleaning up codegen

### LINKED ISSUE #75: Failed guards result in corrupt internal state
During the changes in https://github.com/korken89/smlang-rs/pull/72, we are not properly restoring the state machine state if a guard fails. The state is `take()`n during processing, but is not reset upon guard errors.

### PR DESCRIPTION
**Pull request by ryan-summers**
Fixes #75 by reconfiguring the state if the guard fails. We were unintentionally bailing immediately with no internal state configured due to the `?` operation.

### CODE REVIEW

**Comment by ryan-summers** [comment]
cc @korken89 / @dzimmanck / @pleepleus - I don't think I can review/approve/merge without one of you guys signing off

**Comment by korken89** [comment]
Nice catch! If you get the changelog fixed I'll merge it :)

**Comment by korken89** [comment]
Just for my understanding, why is in-place replacing the state is not used? It would catch these errors at compile-time.

**Comment by ryan-summers** [comment]
It appears that the in-place swap-over of the state to `take()` approach happened in https://github.com/korken89/smlang-rs/pull/49/files#diff-5b70e2aae1cde8f581e01e9a15ac4f570e28b80ca2fb59d2911fd1ac99a30bd9L529. Comments in that PR indicate it was necessary to avoid an `unwrap()` in the code, but I suspect this may not be the case. Let me review to see if we can update it to generate code with in-place management of the state.

**Comment by korken89** [comment]
Ah sorry I thought the bug was existing between releases, no changelog needed :)

I generally use the following create to do in place state management. Though it was not added at the std lib for good reason.
https://crates.io/crates/replace_with

Though this is not the correct PR to change it. 

A closure that works on top of the option would also give the same guarantees and not tred on the limits of what Rust can do. :)

**Comment by ryan-summers** [comment]
Edit: The `Option<State>` change happened because actions were updated to take ownership of internal state data to allow the data to be transferred to a new state.

However, the more I'm thinking about that, the more I think that we should perhaps revert #49. If a user has a large, non-clonable state data, I imagine they could put that in some type of `Box` (or static storage location) and just pass a ref to it inside the various states instead.

Will take a look at the closure based approach though. I agree that #49 has made the internal state pretty fragile. It seems nonsensical that there can be `None` state.

**Comment by korken89** [comment]
Yeah, to me I like replace_with just because it makes the computer enforce that all transitions are correct and you don't need any intermediate representation. Using an option is fine though as well with the closure support, it will give the same result without unsafe. :)

Here is an implementation example:
https://github.com/rust-lang/rfcs/pull/1736#issuecomment-251673355

**Comment by korken89** [comment]
Yeah, you really need to be careful when you can move stuff out. Which way to go is not clear to me, I'm not sure how much work it works be to use replace_with. The revert is the easy option :)

**Comment by ryan-summers** [comment]
Looking at it further, `replace_with` actually has the same draw-back. You get ownership of the internal state data in the `replace_with` closure, but you still lose it if you pass it to `action` and `action` fails, so you can never provide the same `state` data outwards.

I think the only fix here is to remove ownership of internal state data in `action`. The mechanism necessary to move data (without cloning) between states would be to `Box` it (i.e. make an owned ref to it) and clone the `Box` type to the new state.

**Comment by korken89** [comment]
Alright, then the path is set! :)

File: macros/src/lib.rs [hunk 1]
```diff
@@ -65,7 +66,19 @@ pub fn statemachine(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
...
+                name.to_string()
+            } else {
+                "default".to_string()
+            };
+            let mut expansion_file =
+                std::fs::File::create(format!("target/smlang-expansion-{name}.rs")).unwrap();
+            expansion_file
+                .write_all(format!("{tokens}").as_bytes())
+                .unwrap();
+            tokens
```
Comment (ryan-summers) [inline macros/src/lib.rs:81]: I found myself continuously trying to find the expanded code, so I ended up writing this in similar to RTIC. I need to check if this is the write way to get it into the target build directory.

Comment (ryan-summers) [reply macros/src/lib.rs]: Removed this for now. We can add it on a follow-up PR. The target path derivation is more complex than this.

**Comment by ryan-summers** [comment]
Alright, the reversion should be complete here. Thanks for the discussion @korken89 :)

**Pull request merged**